File: jekyun-park/week4/programmers/43164.swift
//
//  main.swift
//  SwiftAlgorithms
//
//  Created by 박제균 on 2022/01/20.
//

import Foundation

func solution(_ tickets: [[String]]) -> [String] {

    var visited = Array(repeating: false, count: tickets.count)
    let tickets = tickets.sorted { $0[1] < $1[1] }
    var path: [String] = []

    func dfs(_ city: String) {
        if path.count == tickets.count {
            path.append(city)
            return
        }

        for i in 0..<tickets.count {
            let start = tickets[i][0]
            let end = tickets[i][1]

            if city == start, visited[i] == false {
                visited[i] = true
                path.append(start)
                dfs(end)

                if path.count == tickets.count + 1 {
                    return
                }

                // 해당 도시에 갈 수 없으면 취소함
                path.removeLast()
                visited[i] = false
            }
        }
    }
    dfs("ICN")
    return path
}

print(solution([["ICN", "SFO"], ["ICN", "ATL"], ["SFO", "ATL"], ["ATL", "ICN"], ["ATL", "SFO"]]))

File: JaeYoung/week04/43164.py
def solution(tickets):
    tickets = sorted(tickets, key=lambda x: (x[0], x[1]))
    records = dict()

    for start, arrive in tickets:
        if start in records:
            records[start].append(arrive)
        else:
            records[start] = [arrive]

    path = []

    stack = ["ICN"]
    while stack:
        top = stack[-1]
        if top not in records or len(records[top]) == 0:
            path.append(stack.pop())
        else:
            stack.append(records[top].pop(0))

    print(path[::-1])

    return path[::-1]
